File: library/src/main/java/com/cz/android/sample/library/component/message/SampleMessageComponent.java
package com.cz.android.sample.library.component.message;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.FragmentActivity;

import com.cz.android.sample.component.ComponentContainer;
import com.cz.android.sample.library.R;
import com.cz.android.sample.library.thread.WorkThread;

import java.util.Observable;
import java.util.Observer;

/**
 * @author Created by cz
 * @date 2020-01-29 16:28
 * @email bingo110@126.com
 */
public class SampleMessageComponent implements ComponentContainer <FragmentActivity>{
    private final WorkThread<String> workThread=new WorkThread("sample_work_thread");
    private final SampleSystemConsole sampleSystemConsole=new SampleSystemConsole();

    public SampleMessageComponent() {
        workThread.startService();
        sampleSystemConsole.setup(workThread);
    }

    @Override
    public boolean isComponentAvailable(@NonNull Object object) {
        SampleMessage sampleMessage = object.getClass().getAnnotation(SampleMessage.class);
        return null!=sampleMessage&&sampleMessage.value();
    }

    @Override
    public View getComponentView(FragmentActivity context, Object object, ViewGroup parentView,View view, Bundle saveInstance) {
        LayoutInflater layoutInflater = LayoutInflater.from(context);
        final View contentLayout = layoutInflater.inflate(R.layout.sample_message_layout, parentView, false);
        FrameLayout sampleMessageContentLayout=contentLayout.findViewById(R.id.sampleMessageContentLayout);
        sampleMessageContentLayout.addView(view);
        return contentLayout;
    }

    @Override
    public void onCreatedView(FragmentActivity context, Object object, View view) {
        //If activity/fragment want to output message
        final NestedScrollView scrollView=view.findViewById(R.id.sampleScrollView);
        final TextView messageView = view.findViewById(R.id.sampleMessageText);
        final Observer observer=new Observer() {
            @Override
            public void update(Observable observable,final Object o) {
                messageView.post(new Runnable() {
                    @Override
                    public void run() {
                        messageView.append(o.toString());
                    }
                });
            }
        };
        final View scrollDownButton=view.findViewById(R.id.scrollDownButton);
        scrollDownButton.setSelected(true);
        final TextWatcher textWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void afterTextChanged(Editable editable) {
            }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                if(scrollDownButton.isSelected()){
                    scrollView.fullScroll(ScrollView.FOCUS_DOWN);
                }
            }
        };
        messageView.addTextChangedListener(textWatcher);
        scrollDownButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                scrollDownButton.setSelected(!scrollDownButton.isSelected());
                if(scrollDownButton.isSelected()){
                    messageView.addTextChangedListener(textWatcher);
                } else {
                    messageView.removeTextChangedListener(textWatcher);
                }
            }
        });
        View clearMessageButton=view.findViewById(R.id.clearMessageButton);
        clearMessageButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                messageView.setText(null);
            }
        });

        workThread.addObserver(observer);
        //here we try to bind SampleMessageBindFragment
        SampleMessageBindFragment.injectIfNeededIn(context,workThread);
    }

    @Override
    public int getComponentPriority() {
        return 1;
    }
}

File: extension/src/main/java/com/github/jackchen/android/sample/library/component/message/SampleMessageComponent.kt
package com.github.jackchen.android.sample.library.component.message

import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.widget.NestedScrollView
import com.github.jackchen.android.core.component.ComponentContainer
import com.github.jackchen.android.sample.api.Extension
import com.github.jackchen.android.sample.library.R
import java.util.Observer

/**
 * @author Created by cz
 * @date 2020-01-29 16:28
 * @email bingo110@126.com
 */
@Extension
class SampleMessageComponent : ComponentContainer {
  private val workThread: WorkThread<String> = WorkThread<String>("sample_work_thread")
  private val sampleSystemConsole = SampleSystemConsole()

  init {
    workThread.startService()
    sampleSystemConsole.setup(workThread)
  }

  override fun isComponentAvailable(component: Any): Boolean {
    val sampleMessage = component.javaClass.getAnnotation(SampleMessage::class.java)
    return null != sampleMessage && sampleMessage.value
  }

  override fun getComponentView(context: AppCompatActivity, component: Any, parentView: ViewGroup, view: View): View {
    val layoutInflater = LayoutInflater.from(context)
    val contentLayout = layoutInflater.inflate(R.layout.sample_message_layout, parentView, false)
    val sampleMessageContentLayout = contentLayout.findViewById<FrameLayout>(R.id.sampleMessageContentLayout)
    sampleMessageContentLayout.addView(view)
    return contentLayout
  }

  override fun onCreatedView(context: AppCompatActivity, `object`: Any, view: View) {
    // If activity/fragment want to output message
    val scrollView = view.findViewById<NestedScrollView>(R.id.sampleScrollView)
    val messageView = view.findViewById<TextView>(R.id.sampleMessageText)
    val observer = Observer { _, o -> messageView.post { messageView.append(o.toString()) } }
    val scrollDownButton = view.findViewById<View>(R.id.scrollDownButton)
    scrollDownButton.isSelected = true
    val textWatcher: TextWatcher = object : TextWatcher {
      override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
      override fun afterTextChanged(editable: Editable) {}
      override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
        if (scrollDownButton.isSelected) {
          scrollView.fullScroll(ScrollView.FOCUS_DOWN)
        }
      }
    }
    messageView.addTextChangedListener(textWatcher)
    scrollDownButton.setOnClickListener {
      scrollDownButton.isSelected = !scrollDownButton.isSelected
      if (scrollDownButton.isSelected) {
        messageView.addTextChangedListener(textWatcher)
      } else {
        messageView.removeTextChangedListener(textWatcher)
      }
    }
    val clearMessageButton = view.findViewById<View>(R.id.clearMessageButton)
    clearMessageButton.setOnClickListener { messageView.text = null }
    workThread.addObserver(observer)
    // here we try to bind SampleMessageBindFragment
    SampleMessageBindFragment.injectIfNeededIn(context, workThread)
  }
}
